Add envWithTpl variable

Th0masL opened the pull request:
Add `envWithTpl` variable that support injecting Environment Variables with supporting values coming from `--set <variable>=<value>`, by using the function `tpl`.

This does not impact the behavior of the Environment Variables defined under `env`, so this is not a breaking change.

Example of values.yaml:
```
mySuperSecret: ""

envWithTpl:
  MY_SUPER_SECRET: "{{ .Values.mySuperSecret }}"
```

Example of helm command:
```
export MY_SUPER_SECRET="password123"

helm upgrade fluent-bit --install --debug --values ./values.yaml --set mySuperSecret=${MY_SUPER_SECRET} fluent/fluent-bit
```

This is a similar fix as https://github.com/fluent/helm-charts/issues/186 that is provided by the PR https://github.com/fluent/helm-charts/pull/202, but uses `range` and does not touch the existing `env` logic.

Th0masL commented:
Do you prefer if I rename the variable as `envWithTpl` ?

stevehipwell commented on charts/fluent-bit/templates/_pod.tpl:
```diff
@@ -37,9 +37,19 @@ containers:
-  {{- if .Values.env }}
+  {{- if or .Values.env .Values.envWithTpl }}
     env:
+      {{- if .Values.env }}
```
Using `with` here would be better than `if`.

Th0masL replied:
Done

stevehipwell commented on charts/fluent-bit/templates/_pod.tpl:
```diff
@@ -37,9 +37,19 @@ containers:
+      {{- if .Values.env }}
       {{- toYaml .Values.env | nindent 6 }}
+      {{- end }}
+      {{- if .Values.envWithTpl }}
+      {{- with .Values.envWithTpl }}
+      {{- range $key, $val := . }}
```
```suggestion
      {{- range $key, $val := . }}
```
You don't need the `if` and `with`.

Th0masL replied:
Indeed, those were not needed, but when I removed them I was unable to get the `range` function working, so I've switched to be simply using `{{- tpl (toYaml .) $ | nindent 6 }}`

stevehipwell commented on charts/fluent-bit/values.yaml:
```diff
@@ -215,6 +215,11 @@ podLabels: {}
+#  - name: FOO
+#    value: "bar"
+
+envWithTpl: {}
```
I'd stick to an array for compatibility.

Th0masL commented:
I applied the suggested improvements, but for some reason I was not able to get the `range` function working when I converted the `envWithTpl` to an array, so I've also removed the `range`.

stevehipwell commented:
> I applied the suggested improvements, but for some reason I was not able to get the range function working when I converted the envWithTpl to an array, so I've also removed the range.

Then this isn't doing anything different to your original submission, I'll put a suggestion in.

stevehipwell commented on charts/fluent-bit/templates/_pod.tpl:
```diff
@@ -37,9 +37,14 @@ containers:
+  {{- if or .Values.env .Values.envWithTpl }}
     env:
-      {{- toYaml .Values.env | nindent 6 }}
+      {{- with .Values.env }}
+      {{- toYaml . | nindent 6 }}
+      {{- end }}
+      {{- with .Values.envWithTpl }}
+      {{- tpl (toYaml .) $ | nindent 6 }}
+      {{- end }}
```
```suggestion
    {{- with .Values.env }}
      {{- toYaml . | nindent 6 }}
    {{- end }}
    {{- range $item := .Values.envWithTpl }}
      - name: {{ $item.name }}
        value: {{ tpl $item.value $ | quote }}
    {{- end }}
```
The code above has ben tested and works correctly.

stevehipwell commented on charts/fluent-bit/values.yaml:
```diff
@@ -215,6 +215,12 @@ podLabels: {}
+#  - name: FOO
+#    value: "bar"
+
+envWithTpl: []
+#  - name: FOO_2
+#    value: "bar2"
```
```suggestion
envWithTpl: []
#  - name: FOO_2
#    value: "{{ .Values.foo2 }}"
#
# foo2: bar2
```
The example would be better showing the intended use. Maybe also add a comment as to why you'd use it over `env`.

Th0masL commented:
Thank you for your suggestions !
I've added a comment to explain the usage of `envWithTpl`.
Let me know if you see any other thing that needs to be improved 🙂

stevehipwell commented:
@Th0masL are you sure that you wouldn't rather use the K8s native way of doing this without modifying the chart (see below)? Your code here is fine but I wouldn't recommend your use case as you always need the credentials present when updating the chart and anyone with Helm access can get the credentials.

Firstly create your secret, this can be in a completely separate one off process, with all of your secret values under their own keys (duplicate the `--from-literal` argument).

```shell
MY_SUPER_SECRET="password"
kubectl create secret generic fluent-bit-secrets --from-literal "my-super-secret=${MY_SUPER_SECRET}"
```

Use the secret to set your pod environment variables.

```yaml
env:
  - name: MY_SUPER_SECRET
    valueFrom:
      secretKeyRef:
        name: fluent-bit-secrets
        key: my-super-secret
```

Th0masL commented:
I understand that some people might prefer to create secrets to do that, but all the other charts that we're using already support passing templatable strings values to Environment Variables, so I would like to keep the same settings also for the fluent-bit chart.

stevehipwell commented:
@Th0masL I would suggest using the K8s native pattern for all of your charts, but I suspect we'll have to agree to disagree.

@naseemkullah what are your thoughts here?

Th0masL commented:
> @Th0masL I would suggest using the K8s native pattern for all of your charts, but I suspect we'll have to agree to disagree.

Yep, I'm fine not agreeing 🙂
My main goal was to find a way to support templatable strings without having to impact the behavior of the existing `env` variable, so it does not break other people's fluent-bit deployments. I think we reached this goal, so I'm happy with the result.

naseemkullah commented:
> @Th0masL I would suggest using the K8s native pattern for all of your charts, but I suspect we'll have to agree to disagree.
> 
> 
> 
> @naseemkullah what are your thoughts here?

While many may mount a secret, Thomas' approach seems reasonable as well.

I'm all for it, though could you remind me what are the dangers/breakage of using templating directly in .Values.env (rather than a new value) are?

stevehipwell commented:
> I'm all for it, though could you remind me what are the dangers/breakage of using templating directly in .Values.env (rather than a new value) are?

The `tpl` command only works correctly when modifying a string with a string so it's best practice to make implementations using it really clear. This pattern, manual field mapping, or testing for a string literal are generally considered the safest way to enable it's use with dynamic inputs.

naseemkullah commented:
thanks @Th0masL , lgtm to me but will wait for @stevehipwell's approval (btw could you try merging when you approve @stevehipwell ? Not sure if we have confirmed you are able to yet)

stevehipwell commented:
@Th0masL you need to bump the version and update the annotations in _Chart.yaml_.

Th0masL commented:
> @Th0masL you need to bump the version and update the annotations in _Chart.yaml_.

Done 🙂

Th0masL commented on charts/fluent-bit/Chart.yaml:
```diff
@@ -23,4 +23,4 @@ maintainers:
 annotations:
   artifacthub.io/changes: |
     - kind: changed
-      description: "Update Fluent Bit image to v1.8.11."
+      description: "Add envWithTpl to support templatable strings in Environment Variables"
```
For some reason it's showing the following error when I try to approve the suggestion from the Web UI.
```
Applying suggestions on deleted lines is currently not supported.
```
So I'll apply your suggestion in a new commit

stevehipwell replied:
I saw similar errors, it looks like a GitHub issue.
